paddle/fluid/operators/diagonal_op.cu

```diff
@@ -0,0 +1,362 @@
+#include "paddle/fluid/platform/cuda_primitives.h"
+
+namespace paddle {
+namespace operators {
+
+using platform::PADDLE_CUDA_NUM_THREADS;
+
+template <typename T, int INPUT_DIM_SIZE, int OUTPUT_DIM_SIZE>
+__global__ void Diagonal(const T* input_data, T* output_data,
+                         const int64_t offset_, int64_t axis1_, int64_t axis2_,
+                         int64_t* input_stride, int64_t* output_stride,
+                         int64_t numel) {
+  CUDA_KERNEL_LOOP(idx, numel) {
+    int64_t idx_dim[INPUT_DIM_SIZE] = {0};
+    int64_t temp = 0;
+    for (size_t i = 0; i < INPUT_DIM_SIZE - 1; i++) {
+      idx_dim[i] = (idx - temp) / input_stride[i];
+      temp = temp + idx_dim[i] * input_stride[i];
+    }
+    idx_dim[INPUT_DIM_SIZE - 1] = idx - temp;
+
+    int64_t axis1_dim = idx_dim[axis1_];
+    int64_t axis2_dim = idx_dim[axis2_];
+
+    int64_t out_dim[OUTPUT_DIM_SIZE] = {0};
+    int temp_pos = 0;
+    for (int i = 0; i < INPUT_DIM_SIZE; i++) {
+      if (i != axis1_ && i != axis2_) {
+        out_dim[temp_pos] = idx_dim[i];
+        temp_pos++;
+      }
+    }
+    bool flag = false;
+    if (offset_ == 0 && axis1_dim == axis2_dim) {
+      out_dim[temp_pos] = axis1_dim;
+      flag = true;
+    } else if (offset_ > 0 && (axis1_dim + offset_) == axis2_dim) {
+      out_dim[temp_pos] = axis1_dim;
+      flag = true;
+    } else if (offset_ < 0 && (axis1_dim + offset_) == axis2_dim) {
+      out_dim[temp_pos] = axis2_dim;
+      flag = true;
+    }
+    if (flag) {
+      int64_t idx_output = 0;
+      for (size_t i = 0; i < OUTPUT_DIM_SIZE - 1; i++) {
+        idx_output = idx_output + out_dim[i] * output_stride[i];
+      }
+      idx_output = idx_output + out_dim[OUTPUT_DIM_SIZE - 1];
+      output_data[idx_output] = input_data[idx];
+    }
+  }
+}
+
+template <typename T>
+class DiagonalCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* input = context.Input<framework::Tensor>("Input");
+    const auto* input_data = input->data<T>();
+    auto input_dim = input->dims().Get();
+    auto input_dim_size = input->dims().size();
+
+    int input_stride_size = input_dim_size - 1;
+    int64_t* host_input_stride;
+    host_input_stride =
+        reinterpret_cast<int64_t*>(malloc(input_stride_size * sizeof(int64_t)));
+    for (size_t i_input = 0; i_input < input_stride_size; i_input++) {
+      int64_t temp_stride = 1;
+      for (size_t j = i_input + 1; j < input_dim_size; j++) {
+        temp_stride = temp_stride * input_dim[j];
+      }
+      host_input_stride[i_input] = temp_stride;
+    }
+    int64_t* input_stride;
+    cudaMalloc(reinterpret_cast<void**>(&input_stride),
+               input_stride_size * sizeof(int64_t));
+    cudaMemcpy(reinterpret_cast<void*>(input_stride),
+               reinterpret_cast<void*>(host_input_stride),
+               input_stride_size * sizeof(int64_t), cudaMemcpyHostToDevice);
```

Comment:
Do not use cudaMemcpy/cudaMalloc in op kernel directly. Firstly, try to use less memory as possible, i.e., do not malloc extra memory if can be avoid by refine the algorithm. Secondly, If  temporary memory is exactly needed, use temp Tensor, and call `tensor.mutable_data`.

Reply:
Replace cudaMalloc/cudaMemcpy by TensorFormVector in cudaKernel and cudaGradKernel.